src/controllers/gamepads.jsx

```diff
@@ -0,0 +1,171 @@
+const MAX_DRIVE_ANGLE = 12;
+const MAX_TRANSLATE_ANGLE = 45;
+
+const MIN_ROTUNDA_ANGLE = -90;
+const MAX_ROTUNDA_ANGLE = 90;
+
+const MIN_SHOULDER_ANGLE = 0;
+const MAX_SHOULDER_ANGLE = 90;
+
+const MIN_ELBOW_ANGLE = -180;
+const MAX_ELBOW_ANGLE = 180;
+
+const MIN_WRIST_PITCH_ANGLE = -90;
+const MAX_WRIST_PITCH_ANGLE = 90;
+
+const MIN_WRIST_ROLL_ANGLE = -360;
+const MAX_WRIST_ROLL_ANGLE = 360;
+
+const MIN_END_EFFECTOR_ANGLE = 0;
+const MAX_END_EFFECTOR_ANGLE = 150;
+
+function clamp(num, min, max) {
+    return Math.min(Math.max(num, min), max);
+}
+
+let driveMapping = {
+    speed : 1,                  // left stick up/down
+    angle : 2,                  // right stick left/right
+    enable_speed : 7,           // right bumper
+    spin_mode : 2,              // X
+    translate_mode : 3,         // Y
+    drive_mode : 1,             // B
+    wheel_orientation_0 : 14,   // D-Pad Left
+    wheel_orientation_1 : 12,   // D-Pad Up
+    wheel_orientation_2 : 15,   // D-Pad Right
+}
+
+let armMapping = {
+    joystickVertical: 1,
+    thumbJoystickVertical: 9,
+    thumbUp: -1,
+    thumbDown: 0.14285719394683838,
+    joystickRotate: 5,
+    joystickTrigger: 0,
+    rudder: 6,
+    button3: 2,
+    button4: 3,
+    button5: 4,
+    button6: 5,
+}
+
+class DriveGamepad {
+    constructor(gamepad) {
+        this.gamepad = gamepad;
+    }
+
+    getMode(currentMode) {
+        let mode = currentMode;
+        const spin_mode = this.gamepad.buttons[2].pressed;
+        const translate_mode = this.gamepad.buttons[3].pressed;
+        const drive_mode = this.gamepad.buttons[1].pressed;
+
+        if (spin_mode) {
+            mode = "spin"
+        } else if (translate_mode) {
+            mode = "translate"
+        } else if (drive_mode) {
+            mode = "drive"
+        } else {
+            mode = currentMode
+        }
+        return mode;
+    }
+    
+    getSpeed(currentSpeed) {
+        const throttleSpeed = parseInt((this.gamepad?.axes[driveMapping.speed] * -10).toFixed(0)) * 5;
```

Comment:
A while ago I was experimenting with non-linear speed functions for mapping controller inputs to system outputs. I think something like this might feel better to drive, but I never actually managed to test it out. Also a smooth curve would probably be better than literally this, but I'm just trying to illustrate the idea. Regardless, this looks great.

![image](https://github.com/user-attachments/assets/b68f1b48-3432-4a12-bb62-807e13a0aa31)